Logic/item_detection.py

```diff
@@ -0,0 +1,196 @@
+
+    def find_ball_parameters(self, image_path, hole_positions):
+        '''Iterate through a range of values and saving each result in the goal of identify the best parameters for ball detection'''
+
+        min_x = min(hole_positions, key=lambda t: t[0])[0] + 10
+        min_y = min(hole_positions, key=lambda t: t[1])[1] + 10
+        max_x = max(hole_positions, key=lambda t: t[0])[0] - 10
+        max_y = max(hole_positions, key=lambda t: t[1])[1] - 10
+
+        for param2 in range(13, 17, 1):
+            for param1 in range(10, 310, 10):
+                print('Ball Parameters:', param1, param2)
+
+                marker_image = cv2.imread(self.lookup.marker)
+                marker_image = cv2.cvtColor(marker_image, cv2.COLOR_BGR2GRAY)
+
+                rgb_image = cv2.imread(image_path)
+                gray_image = cv2.cvtColor(rgb_image, cv2.COLOR_BGR2GRAY)
+
+                circles = cv2.HoughCircles(gray_image, cv2.HOUGH_GRADIENT, 1, 10, param1=param1, param2=param2, minRadius=9, maxRadius=13)
+
+                if circles is not None:
+                    circles = np.round(circles[0, :]).astype("int")
+
+                    for (x_position, y_position, radius) in circles:
+                        radius = 13 # assuming ball is always 13 pixels
+
+                        if min_x <= x_position and x_position <= max_x and min_y <= y_position and y_position < max_y:
+                            check_image = gray_image[(y_position - radius) : (y_position + radius), (x_position - radius) : (x_position + radius)]
+
+                            if ssim(check_image, marker_image) < 0:
+                                cv2.circle(rgb_image, (x_position, y_position), 2, (15, 10, 25), 3)
+                                cv2.circle(rgb_image, (x_position, y_position), radius, (25, 10, 255), 3)
+
+                cv2.imwrite(self.lookup.ball_training_path + str(param2) + '_' + str(param1) + '.jpg', rgb_image)
+
+    def detect_holes(self, frame):
+        '''Responsible for returning an array of hole positions'''
+
+        detected_holes = []
+
+        gray_image = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
+
+        holes = cv2.HoughCircles(gray_image, cv2.HOUGH_GRADIENT, 1, 20, param1=150, param2=16, minRadius=20, maxRadius=21)
+
+        if holes is not None:
+            holes = np.round(holes[0, :]).astype("int")
+
+            for (x_position, y_position, radius) in holes:
+                detected_holes.append((x_position, y_position))
+
+        return detected_holes
+
+    def detect_balls(self, frame, hole_positions):
+        '''Responsible for returning an array of ball positions'''
+
+        detected_circles = []
+
+        min_x = min(hole_positions, key=lambda t: t[0])[0] + 10
+        min_y = min(hole_positions, key=lambda t: t[1])[1] + 10
+        max_x = max(hole_positions, key=lambda t: t[0])[0] - 10
+        max_y = max(hole_positions, key=lambda t: t[1])[1] - 10
```

Comment:
Refactor, make a function that gives the 4 corners in a list and call it

Reply:
Done